tritium-registry/src/main/java/com/palantir/tritium/metrics/registry/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -132,15 +132,19 @@ private void addSample(double priority, long value, double itemWeight, boolean b
         State rescale(long newTick) {
             long durationNanos = newTick - startTick;
             double scalingFactor = Math.exp(-alphaNanos * durationNanos);
-            final int newCount;
+            int newCount = 0;
             ConcurrentSkipListMap<Double, WeightedSample> newValues = new ConcurrentSkipListMap<>();
             if (Double.compare(scalingFactor, 0) != 0) {
                 RescalingConsumer consumer = new RescalingConsumer(scalingFactor, newValues);
                 values.forEach(consumer);
                 // make sure the counter is in sync with the number of stored samples.
                 newCount = consumer.count;
-            } else {
-                newCount = 0;
+            }
+            // It's possible that more values were added while the map was scanned, those with the
+            // minimum priorities are removed.
+            while (newCount > size) {
+                newValues.pollFirstEntry();
+                newCount--;
```

Comment:
pedantic nit, I think we're guaranteed that the values won't be null since newValues won't be empty, but to make it clear, would suggest the following

```suggestion
                if (newValues.pollFirstEntry() != null) {
                    newCount--;
                }
```

Reply:
:+1:

Reply:
Perhaps we should throw in the null case, otherwise we might spin indefinitely.